core/shared/src/main/scala/org/http4s/Message.scala

```diff
@@ -465,6 +465,22 @@ final class Request[+F[_]] private (
   def isIdempotent: Boolean =
     method.isIdempotent || headers.contains[`Idempotency-Key`]
 
+  /**
+   * Checks whether the [[Entity]] of the [[Message]] is pure in the sense of
+   * indicating that the entity body can be re-run without side effects.
+   *
+   * @return the [[Boolean]] value:
+   *         'true' if entity is empty or already loaded into memory;
+   *         'false' if entity is streamed.
+   */
+  def isPureEntity: Boolean =
```

Comment:
Probably, we could add this to the `isIdempotent` method directly. But I feel it's too risky because it may hurt people's runtime in bizarre ways. Changing the behavior of the `Retry` middleware seems a bit less risky. I could be wrong here, so your comments are appreciated so much.

Reply:
I could be wrong about this, but I think there's an interesting distinction here between an idempotent request (i.e., has no side-effect on the server-side) and a request with a pure entity (i.e., has no side-effect in the client-side).